airbyte-commons-worker/src/main/java/io/airbyte/workers/internal/HeartbeatTimeoutChaperone.java

```diff
@@ -0,0 +1,130 @@
+
+  private final HeartbeatMonitor heartbeatMonitor;
+  private final Duration timeoutCheckDuration;
+  private final FeatureFlagClient featureFlagClient;
+  private final UUID workspaceId;
+
+  private boolean hasFailed = false;
+
+  public HeartbeatTimeoutChaperone(final HeartbeatMonitor heartbeatMonitor,
+                                   final Duration timeoutCheckDuration,
+                                   final FeatureFlagClient featureFlagClient,
+                                   final UUID workspaceId) {
+    this.timeoutCheckDuration = timeoutCheckDuration;
+
+    LOGGER.info("Starting source heartbeat check. Will check every {} minutes.", timeoutCheckDuration.toMinutes());
+
+    this.heartbeatMonitor = heartbeatMonitor;
+    this.featureFlagClient = featureFlagClient;
+    this.workspaceId = workspaceId;
+  }
+
+  public Runnable runWithHeartbeatThread(final Runnable runnable, final ExecutorService executorService) {
+    return () -> {
+      final CompletableFuture<Void> runnableFuture = CompletableFuture.runAsync(runnable, executorService);
+      final CompletableFuture<Void> heartbeatFuture = CompletableFuture.runAsync(this::monitor, executorService);
+
+      try {
+        CompletableFuture.anyOf(runnableFuture, heartbeatFuture).get();
+      } catch (final InterruptedException e) {
+        LOGGER.error("Heartbeat chaperone thread was interrupted.", e);
+        return;
+      } catch (final ExecutionException e) {
+        // this should check explicitly for source and destination exceptions
+        if (e.getCause() instanceof RuntimeException) {
+          throw (RuntimeException) e.getCause();
+        } else {
+          throw new RuntimeException(e);
+        }
+      }
+
+      LOGGER.info("thread status... heartbeat thread: {} , replication thread: {}", heartbeatFuture.isDone(), runnableFuture.isDone());
+
+      boolean heartbeatTimedOut = false;
+      if (heartbeatFuture.isDone() && !runnableFuture.isDone()) {
+        heartbeatTimedOut = true;
+        runnableFuture.cancel(true);
+      }
+
+      if (runnableFuture.isDone() && !heartbeatFuture.isDone()) {
+        heartbeatFuture.cancel(true);
+      }
+
+      if (heartbeatTimedOut) {
+        throw new HeartbeatTimeoutException(String.format("Heartbeat has stopped. Heartbeat freshness threshold: %s Actual heartbeat age: %s",
+            heartbeatMonitor.getHeartbeatFreshnessThreshold(),
+            heartbeatMonitor.getTimeSinceLastBeat().orElse(null)));
+      }
+    };
+  }
+
+  @SuppressWarnings("BusyWait")
+  private void monitor() {
+    while (true) {
+      try {
+        sleep(timeoutCheckDuration.toMillis());
+      } catch (final InterruptedException e) {
+        LOGGER.info("Heartbeat thread has been interrupted (this is expected; the heartbeat was healthy the whole time).");
+        return;
+      }
+
+      // if not beating, return. otherwise, if it is beating or heartbeat hasn't started, continue.
```

Comment:
Is it worth moving the "heartbeat hasn't started" to another loop before this one? I think we may eventually have a config on how much time do we give to a connector before the first message.

Reply:
I am not sure if it worth it since now we're using the logs as a beat it seems to be unlikely to not have a log message. Could we test it first? I may add a datadog monitor for it if it helps.